File: packages/agent/src/db/memory.ts
import { eq, sql } from "drizzle-orm";
import type { AgentDatabase } from "./client";
import { memory } from "./schema";

// Agent-side memory KV. Lives in the `memory` table of `agent.db`. This is the
// freeform store for anything the agent wants to remember between sessions
// that doesn't fit a typed table — watermarks, last-seen markers, small
// notes. Distinct from MCP-side `tokens.db`, which holds integration
// state (OAuth tokens, queues, caches) the MCP process owns.

export interface MemoryStore {
  get(key: string): string | null;
  set(key: string, value: string): void;
}

export function createMemoryStore(db: AgentDatabase): MemoryStore {
  return {
    get(key) {
      const row = db
        .select({ value: memory.value })
        .from(memory)
        .where(eq(memory.key, key))
        .get();
      return row?.value ?? null;
    },
    set(key, value) {
      db.insert(memory)
        .values({ key, value })
        .onConflictDoUpdate({
          target: memory.key,
          set: { value, updatedAt: sql`(datetime('now'))` },
        })
        .run();
    },
  };
}

// Well-known keys injected into the session context block. Keep them here
// so writers and the supervisor agree on naming.
export const MEMORY_KEYS = {
  newsLastReadAt: "news_digest.last_read_at",
} as const;

File: agent-py/agent_py/db/memory.py
"""Agent-side memory KV.

A small key-value store over sqlite. Freeform storage for anything the agent
wants to remember between sessions that does not fit a typed table: watermarks,
last-seen markers, notes.
"""

from __future__ import annotations

import os
import sqlite3
from pathlib import Path

_DEFAULT_PATH = Path(__file__).resolve().parent.parent.parent / "data" / "agent.db"

_conn: sqlite3.Connection | None = None


def _db() -> sqlite3.Connection:
    global _conn
    if _conn is None:
        path = os.environ.get("AGENT_DB_PATH", str(_DEFAULT_PATH))
        Path(path).parent.mkdir(parents=True, exist_ok=True)
        conn = sqlite3.connect(path, check_same_thread=False)
        conn.execute("PRAGMA journal_mode = WAL")
        conn.execute(
            """CREATE TABLE IF NOT EXISTS memory (
                 key        TEXT PRIMARY KEY,
                 value      TEXT NOT NULL,
                 updated_at TEXT NOT NULL DEFAULT (datetime('now'))
               )"""
        )
        conn.commit()
        _conn = conn
    return _conn


def get_memory(key: str) -> str | None:
    row = _db().execute("SELECT value FROM memory WHERE key = ?", (key,)).fetchone()
    return row[0] if row else None


def set_memory(key: str, value: str) -> None:
    _db().execute(
        """INSERT INTO memory (key, value) VALUES (?, ?)
           ON CONFLICT(key) DO UPDATE SET value = excluded.value,
                                          updated_at = datetime('now')""",
        (key, value),
    )
    _db().commit()


# Well-known keys injected into the session context block. Kept here so writers
# and the supervisor agree on naming.
MEMORY_KEYS = {
    "news_last_read_at": "news_digest.last_read_at",
}
